### PR TITLE
[WIP] Few mn list sync changes

### PR DESCRIPTION
**Pull request by UdjinM6**
Changes:
- add more mn states
- only remove spent
- send mn ping in addition to mn announce on sync
- manage mn announces more carefully

This also adjusts active mn auto-start logic accordingly.

### CODE REVIEW

**Comment by UdjinM6** [comment]
Missed one piece.
```
-        bool fWatchdogActive = mnodeman.IsWatchdogActive();
+        bool fWatchdogActive = masternodeSync.IsSynced() && mnodeman.IsWatchdogActive();
```
Also, note that since currently nodes do not send pings on sync you can't really test this (unless you connect to you own upgraded node) and starting with old mncache.dat will probably make nodes look expired. And when you are removing mncache.dat make sure to remove governance.dat too.

**Comment by tgflynn** [comment]
I have been trying to test this but right now I am having problems getting a single node synced.

There seems to be a lot of forking/network fragmentation at the moment.

**Comment by tgflynn** [comment]
This does not appear to fix the problem of masternodes not propagating.

I ran this patch on 2 nodes with the second node having only a single connection to the first node and clearing mncache.dat, mnpayments.dat and governance.dat.

One masternode did not propagate to the second node.

The error message was:

2016-12-23 01:19:12     dash-msghand | MNANNOUNCE -- Masternode announce, masternode=06069e73139bbcfabb9371c8683190566988bf019deaeb090413f2bee0786613-1
2016-12-23 01:19:12     dash-msghand | CMasternodeMan::CheckMnbAndUpdateMasternodeList -- masternode=06069e73139bbcfabb9371c8683190566988bf019deaeb090413f2bee0786613-1
2016-12-23 01:19:12     dash-msghand | CMasternodeMan::CheckMnbAndUpdateMasternodeList -- masternode=06069e73139bbcfabb9371c8683190566988bf019deaeb090413f2bee0786613-1 new
2016-12-23 01:19:12     dash-msghand | CMasternodePing::SimpleCheck -- Masternode ping is invalid, unknown block hash: masternode=06069e73139bbcfabb9371c8683190566988bf019deaeb090413f2bee0786613-1 blockHash=000000855014ce3ce58dcd3e0d18fdf715cb09c14cc5ea0ed62cad89ed3a4c4d
2016-12-23 01:19:12     dash-msghand | CMasternodeMan::CheckMnbAndUpdateMasternodeList -- SimpleCheck() failed, masternode=06069e73139bbcfabb9371c8683190566988bf019deaeb090413f2bee0786613-1

The cases I have seen of failed MN propagation are due to CMasternodePing::SimpleCheck() failing.

I think this function should not be run when adding masternodes from broadcasts (ie. these lines:

https://github.com/dashpay/dash/blob/v0.12.1.x/src/masternode.cpp#L492

should be removed).

**Comment by tgflynn** [comment]
What is the difference in behavior between MASTERNODE_EXPIRED and MASTERNODE_NEW_START_REQUIRED ?

**Comment by UdjinM6** [comment]
> 2016-12-23 01:19:12 dash-msghand | CMasternodePing::SimpleCheck -- Masternode ping is invalid, unknown block hash: masternode=06069e73139bbcfabb9371c8683190566988bf019deaeb090413f2bee0786613-1 blockHash=000000855014ce3ce58dcd3e0d18fdf715cb09c14cc5ea0ed62cad89ed3a4c4d

This is fine, there is no such blockhash on testnet https://test.explorer.dash.org/search?q=000000855014ce3ce58dcd3e0d18fdf715cb09c14cc5ea0ed62cad89ed3a4c4d so ping is actually invalid

> What is the difference in behavior between MASTERNODE_EXPIRED and MASTERNODE_NEW_START_REQUIRED ?

The logic now is that you can recover from any other expired state but not from MASTERNODE_NEW_START_REQUIRED (which is set to 3h).

**Comment by tgflynn** [comment]
It's not fine because it allows a discrepancy to exist between MN lists on different nodes which can lead to forks.  Just because some node is on the wrong chain doesn't prevent it from creating governance objects.

**Comment by UdjinM6** [comment]
> Just because some node is on the wrong chain doesn't prevent it from creating governance objects.

Then we probably should not accept governance objects from forked masternodes. Why would we need smth from a forked mn?

**Comment by tgflynn** [comment]
What if it creates an object or a vote before forking ?

**Comment by UdjinM6** [comment]
I guess that without proper ping it should expire/require new start -> this should invalidate gov object/votes created by such mn -> some another mn/sentinel should figure it out and create a new one if needed. Smth like this?

**Comment by tgflynn** [comment]
As we've discussed before validity of governance objects shouldn't depend on MN state.  As long as the collateral transaction is still valid objects should remain valid.

**Comment by tgflynn** [comment]
If the ping is invalid the mn can be put in the expired state, but it should not be prevented from being added to the list.

**Comment by UdjinM6** [comment]
ok, makes sense, will update this part

**Comment by tgflynn** [comment]
Retesting now.

**Comment by UdjinM6** [comment]
Thanks for testing! 👍 

I added compatibility for old nodes (same logic for `fWaitForPing` as it was before this PR, <70204 nodes only), see f6411e2. This shouldn't affect testnet but should help to avoid confusion on mainnet (otherwise new wallet would see old legit nodes as MASTERNODE_NEW_START_REQUIRED, which is no good imo).

File: src/masternode.cpp [hunk 1]
```diff
@@ -229,6 +229,22 @@ void CMasternode::Check(bool fForce)
     // keep old masternodes on start, give them a chance to receive updates...
     bool fWaitForPing = !masternodeSync.IsMasternodeListSynced() && !IsPingedWithin(MASTERNODE_MIN_MNP_SECONDS);
 
+    //
+    // REMOVE AFTER MIGRATION TO 12.1
+    //
+    // Old nodes do send pings on dseg, so they could switch to one of the expired states
```
Comment (tgflynn) [inline src/masternode.cpp]: Is there a typo here ?  Shouldn't it be "old nodes do not send pings on dseg" ?

Comment (UdjinM6) [inline src/masternode.cpp]: oops, fixed :)

**Pull request merged**